File: csharp/FountainTools.Tests/WriterTest.cs
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2024 Ian Thomas

namespace FountainTools.Tests;
using System.IO;
using Fountain;

public class WriterTest
{
    private string loadTestFile(string fileName)
    {
        return File.ReadAllText("../../../../../tests/" + fileName);
    }

    [Fact]
    public void TestWrite()
    {
        string match = loadTestFile("Writer-output.fountain");

        Parser fp = new Parser();
        fp.AddText(loadTestFile("TitlePage.fountain"));
        fp.AddText(loadTestFile("Sections.fountain"));
        fp.AddText(loadTestFile("Character.fountain"));
        fp.AddText(loadTestFile("Dialogue.fountain"));

        Writer fw = new Writer();
        string output = fw.Write(fp.Script);

        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void TestWriteUTF8()
    {
        string match = loadTestFile("UTF8-output.fountain");

        Parser fp = new Parser();
        fp.AddText(loadTestFile("UTF8.fountain"));

        Writer fw = new Writer();
        string output = fw.Write(fp.Script);

        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }
    
    [Fact]
    public void TestAppend()
    {
        string match = loadTestFile("Append-output.fountain");

        Script script = new Script();
        script.AddElement(new Character("FRED"));
        script.AddElement(new Dialogue("Test dialogue."));
        script.AddElement(new Character("FRED"), true);
        script.AddElement(new Dialogue("Test dialogue 2."));

        script.AddElement(new Action("Test action 1."));
        script.AddElement(new Action("Test action 2."), true);
        script.AddElement(new Action("Test action 3."));

        script.AddElement(new Character("FRED"));
        script.AddElement(new Dialogue("Test dialogue."));
        script.AddElement(new Character("FRED"));
        script.AddElement(new Dialogue("Test dialogue 2."));

        Writer fw = new Writer();
        string output = fw.Write(script);

        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }
}

File: javascript/test/fountain/writer.test.js
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2024 Ian Thomas

import { loadTestFile } from '../../test/testUtils.js';
import { strict as assert } from 'assert';
import { FountainWriter } from '../../src/fountain/writer.js';
import { FountainParser } from '../../src/fountain/parser.js';
import { Script, Action, Character, Dialogue } from '../../src/screenplay.js';

describe('Writer', () => {

    describe('Simple Parse and Write', () => {
        it('should match', () => {

            const match = loadTestFile('Writer-output.fountain');

            let fp = new FountainParser();

            fp.addText(loadTestFile('TitlePage.fountain'));
            fp.addText(loadTestFile('Sections.fountain'));
            fp.addText(loadTestFile('Character.fountain'));
            fp.addText(loadTestFile('Dialogue.fountain'));

            let fw = new FountainWriter();
            let output = fw.write(fp.script);

            //console.log(output);
            assert.equal(output, match);
            return;
        });
    });

    describe('Write UTF8', () => {
        it('should match', () => {

            const match = loadTestFile('UTF8-output.fountain');

            let fp = new FountainParser();

            fp.addText(loadTestFile('UTF8.fountain'));

            let fw = new FountainWriter();
            let output = fw.write(fp.script);

            //console.log(output);
            assert.equal(output, match);
            return;
        });
    });

    describe('Append', () => {
        it('should match', () => {

            const match = loadTestFile('Append-output.fountain');

            const script = new Script();
            script.addElement(new Character("FRED"));
            script.addElement(new Dialogue("Test dialogue."));
            script.addElement(new Character("FRED"), true);
            script.addElement(new Dialogue("Test dialogue 2."));

            script.addElement(new Action("Test action 1."));
            script.addElement(new Action("Test action 2."), true);
            script.addElement(new Action("Test action 3."));

            script.addElement(new Character("FRED"));
            script.addElement(new Dialogue("Test dialogue."));
            script.addElement(new Character("FRED"), false);
            script.addElement(new Dialogue("Test dialogue 2."));

            let fw = new FountainWriter();

            let output = fw.write(script);

            //console.log(output);
            assert.equal(output, match);
            return;
        });
    });

});